collect/central_collector.go

```diff
@@ -119,18 +153,235 @@ func (c *CentralCollector) collect() {
 
 }
 
-func (c *CentralCollector) processSpan(sp *types.Span) {
-	// add the span to the cache
-	trace := c.SpanCache.Get(sp.TraceID)
-	if trace == nil {
-		err := c.SpanCache.Set(sp)
-		if err != nil {
-			c.Logger.Error().Logf("error adding span with trace ID %s to cache: %s", sp.TraceID, err)
-			return
+func (c *CentralCollector) processTraces() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			ids := c.SpanCache.GetTraceIDs(processTracesBatchSize)
+			if len(ids) == 0 {
+				continue
+			}
+
+			statuses, err := c.Store.GetStatusForTraces(context.Background(), ids)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			for _, status := range statuses {
+				switch status.State {
+				case centralstore.DecisionKeep:
+					c.send(status)
+					c.SpanCache.Remove(status.TraceID)
+
+				case centralstore.DecisionDrop:
+					c.SpanCache.Remove(status.TraceID)
+				default:
+					c.Logger.Debug().Logf("trace %s is still pending", status.TraceID)
+				}
+			}
+		}
+
+		select {
```

Comment:
I feel like everything from here down could be replaced by `time.Sleep(processTracesBackoffInterval)` since the interval is so short. But if we want this, I think we can use a Timer, not a Ticker (since we want it to fire only once) and skip this early test for c.done. It doesn't win us anything that I can tell.

The same is true in `decide()`.

Reply:
Good catch! A Timer is better for this. 
I was thinking a timer allows us to also monitor shutdown signal here without having to wait for the duration we set for the timer.

Reply:
The timer shouldn't be named ticker anymore, and you still don't need the lines at 190-194.